instance/tests/models/test_server.py

```diff
@@ -125,51 +125,117 @@ def authenticate(client):
+
+    @patch('instance.models.server.OpenStackServer.update_status')
+    @patch('instance.models.server.time.sleep')
+    def test_sleep_until_state_changes(self, mock_sleep, mock_update_status):
+        """
+        Check if sleep_until behaves correctly if condition to wait for
+        is unfulfilled initially.
+        """
+        conditions = [
+            lambda: server.status.is_steady_state,
+            lambda: server.status.accepts_ssh_commands,
         ]
-        status_queue.reverse() # To be able to use pop()
+
+        def scoped_update_status(server=None, status_queue=None):
+            """ Return mock update_status scoped to a specific server and status_queue """
+            def update_status():
+                """ Simulate status progression """
+                server._transition(status_queue.pop(), progress=ServerProgress.Success)
+            return update_status
+
+        for condition in conditions:
+            server = OpenStackServerFactory()
+            status_queue = [
+                server._status_to_started,
+                server._status_to_active,
+                server._status_to_booted,
+            ]
+            status_queue.reverse() # To be able to use pop()
+
+            mock_update_status.side_effect = scoped_update_status(server, status_queue)
+            mock_sleep.call_count = 0
+
+            # Sleep until condition is fulfilled
+            server.sleep_until(condition)
+            self.assertEqual(server.status, ServerStatus.Booted)
+            self.assertEqual(server.progress, ServerProgress.Success)
+            self.assertEqual(mock_sleep.call_count, 2)
+
+    @patch('instance.models.server.OpenStackServer.update_status')
+    def test_sleep_until_steady_state(self, mock_update_status):
+        """
+        Check if sleep_until behaves correctly if condition to wait for
+        can not be fulfilled because server is in a steady state
+        (that doesn't fulfill the condition).
+        """
+        server = OpenStackServerFactory()
 
         def update_status():
-            """ Simulate status progression successive runs """
-            new_status = status_queue.pop()
-            if new_status:
-                server._transition(new_status, progress=ServerProgress.Success)
+            """ Simulate status progression """
+            server._transition(server._status_to_started, progress=ServerProgress.Success)
         mock_update_status.side_effect = update_status
 
-        self.assertEqual(server.sleep_until_status(ServerStatus.Booted), ServerStatus.Booted)
-        self.assertEqual(server.progress, ServerProgress.Success)
-        self.assertEqual(server.status, ServerStatus.Booted)
-        self.assertEqual(mock_sleep.call_count, 3)
-        self.assertEqual(status_queue, [server._status_to_terminated])
+        # Pretend that Status.Started (which doesn't accept SSH commands) is a steady state
+        ServerStatus.Started.is_steady_state = True
```

Comment:
@itsjeyd Shouldn't this line affect all the test running after the test? COnsider wrapping everything in catfch-all exception handler and restoring old value, or just using `mock.patch`